lib/runtime/modules.coffee

```diff
@@ -1,130 +1,153 @@
+      main
+    else
+      "#{main}.#{sub}"
+
+  # Choosing Modules
+
+  itemSelector: 'atom-text-editor[data-grammar="source julia"], ink-console.julia'
+
+  isValidItem: (item) -> atom.views.getView(item)?.matches @itemSelector
+
+  autodetect: 'Auto Detect'
+  follow: 'Follow Editor'
+
+  chooseModule: ->
+    item = atom.workspace.getActivePaneItem()
+    ised = atom.workspace.isTextEditor item
+    return unless @isValidItem item
+    client.require =>
+      if (item = atom.workspace.getActivePaneItem())
+        modules = allmodules().then (modules) =>
+          modules.unshift @autodetect if ised
+          modules.unshift @follow if not ised
+          modules
+        selector.show(modules).then (mod) =>
+          return unless mod?
+          if mod is @autodetect
+            delete item.juliaModule
+          else
+            item.juliaModule = mod
+          @updateForItem item
+
+  updateForItem: (item = atom.workspace.getActivePaneItem()) ->
+    @itemSubs.dispose()
+    if not @isValidItem item
+      @itemSubs.add item?.onDidChangeGrammar? => @updateForItem()
+      @setCurrent()
+    else if not client.isConnected()
+      @setCurrent main: 'Main', inactive: true
+    else if atom.workspace.isTextEditor item
+      @updateForEditor item
+    else
+      @setCurrent main: item.juliaModule or 'Main'
+
+  updateForEditor: (editor) ->
+    @setCurrent main: editor.juliaModule or 'Main', true
+    @getEditorModule editor
+    @itemSubs.add editor.onDidChangeCursorPosition =>
+      @getEditorModuleLazy editor
+
+  getEditorModule: (ed) ->
+    return unless client.isConnected()
+    {row, column} = ed.getCursors()[0].getBufferPosition()
     data =
       path: ed.getPath()
       code: ed.getText()
       row: row+1, column: column+1
       module: ed.juliaModule
 
-    client.rpc('module', data).then ({main, sub, inactive, subInactive}) =>
-      @reset main, sub
-      subInactive && @subInactive()
-      inactive && @inactive()
+    getmodule(data).then (mod) =>
+      if atom.workspace.getActivePaneItem() is ed
+        @setCurrent mod, true
 
-  # TODO: auto detect option, remove reset command
-  chooseModule: ->
-    client.require =>
-      selector.show(client.rpc('allmodules')).then (mod) =>
-        return unless mod?
-        atom.workspace.getActiveTextEditor().juliaModule = mod
-        @update()
+  getEditorModuleLazy: debounce ((ed) -> @getEditorModule(ed)), 100
 
-  resetModule: ->
-    client.require =>
-      delete atom.workspace.getActiveTextEditor().juliaModule
-      @update()
+  # The View
+
+  activateView: ->
+    @onDidChange (c) => @updateView c
+
+    @dom = document.createElement 'span'
+    @dom.classList.add 'julia', 'inline-block'
+
+    @mainView = document.createElement 'a'
+    @dividerView = document.createElement 'span'
+    @subView = document.createElement 'span'
+
+    @dom.appendChild x for x in [@mainView, @dividerView, @subView]
+
+    @mainView.onclick = =>
```

Comment:
The idea is that what's choosable == what's clickable. The main module is the bit that you actually choose, and it has no effect on the submodule since that's determined by the code in the buffer itself.